### PR TITLE
Add: create/get credentials in oauth service

### PR DESCRIPTION
**Pull request by Fraggle**
## Description

Add the ability to create and retrieve some credentials from the oauth service.
I moved the seal / unseal methods to a encryption module in order to encrypt the credentials at rest.

## Risk

My first ever rust code.
(tested locally with curl)

## Deploy Plan

Deploy `oauth`

### CODE REVIEW

**Review by spolu — DISMISSED** [review]
Looks great! 

Let's make sure to have a second pair of eyes before merging :+1:

File: core/src/oauth/credential.rs
Comment (fontanierh) [inline core/src/oauth/credential.rs:13]: I don't think we want that

**Review by fontanierh — COMMENTED** [review]
Overall looks great but IMO we should get rid of "credentials provider". Oauth shouldn't care / know about this

File: core/src/oauth/credential.rs
Comment (fontanierh) [reply core/src/oauth/credential.rs:13]: If we want to add a new DB type we shouldn't need to touch `oauth` code at all I think

Comment (Fraggle) [reply core/src/oauth/credential.rs:13]: Considering everything else we'll probably need to do to support another db, i think it's not a big deal.
I personally feel that there is some value to be able to see what each row is, especially as the content of the json is encrypted.
Do you have strong opinions on this or can I merge like that ?

Comment (fontanierh) [reply core/src/oauth/credential.rs:13]: > Considering everything else we'll probably need to do to support another db, i think it's not a compelling argument.

In my view, not having to deploy an extra service to add an enum value we don't use anywhere is a good enough reason not to add this.

I guess if you really want to I can live with it, but I don't see the value in making this a required part of the API (the credentials are opaque anyway and we don't really "look" at the rows). In my opinion, less is more when it comes to APIs.

Feel free to merge as is of course, you already have an approval

File: core/bin/oauth.rs
Comment (fontanierh) [inline core/bin/oauth.rs:210]: I don't think this should mention "json" at all. IMO it's just a string, the user of this API may decide to use this to store JSON strings, but I don't think this should be requirement (you could store private keys or other non-JSON secrets)

File: core/bin/oauth.rs
Comment (fontanierh) [inline core/bin/oauth.rs:220]: same here, I would just make this param a string

Comment (fontanierh) [reply core/bin/oauth.rs:220]: Not sure why it is being wrapped in a serde string ?

Comment (Fraggle) [reply core/bin/oauth.rs:220]: > Not sure why it is being wrapped in a serde string ?

I don't really know either :)

File: core/bin/oauth.rs
Comment (Fraggle) [reply core/bin/oauth.rs:210]: I see your point, will change.

We are very close to turn this into a key/value store :)
Why not just use a key/value store btw ?

Comment (fontanierh) [reply core/bin/oauth.rs:210]: It really is an encrypted key value store i think. We could use a premade one, but imo cost of adding another piece of infra and code to interface with it is higher than adding a couple of endpoints to oauth which we already use for other connectors. Wdyt?

File: core/src/oauth/credential.rs
Comment (Fraggle) [reply core/src/oauth/credential.rs:13]: I'll keep it, to give some intent for it, as well as potential future uses (analytics, debug etc..).
Hopefully you'll be able to live with it :)

Comment (fontanierh) [reply core/src/oauth/credential.rs:13]: Def wont lose sleep over it

Comment (fontanierh) [reply core/src/oauth/credential.rs:13]: (Note that we’ll be able to debug the same thing easily anyway since connectors db will hold both secret id and connector provider)

File: core/bin/oauth.rs
Comment (spolu) [reply core/bin/oauth.rs:220]: We make sure it's JSON but then consider it as a string and store it as such. If aligned with Connection logic all good 👍

File: core/bin/oauth.rs
Comment (spolu) [reply core/bin/oauth.rs:210]: It's an interesting question.

I think it's more extensible to store a JSON here no?

Will make a friendlier API

Comment (fontanierh) [reply core/bin/oauth.rs:210]: The other way around no? If you just store a string, you can store json but also store non-json stuff

Comment (fontanierh) [reply core/bin/oauth.rs:210]: We should do the same as other “secret” managers do IMO (k8s or gcp just store a string)

File: core/bin/oauth.rs
Comment (fontanierh) [reply core/bin/oauth.rs:220]: doesn’t help much to wrap in in a serde string though. Any string can be a serde string so it doesnt validate anything

File: core/bin/oauth.rs
Comment (spolu) [reply core/bin/oauth.rs:210]: I'm neutral

Comment (fontanierh) [reply core/bin/oauth.rs:210]: One argument in favor of enforcing json objects is that it removes the trisk of storing strings that have multiple levels of json.stringify by mistake

File: core/bin/oauth.rs
Comment (spolu) [reply core/bin/oauth.rs:220]: Yes this line makes no sense indeed.

File: core/bin/oauth.rs
Comment (spolu) [reply core/bin/oauth.rs:210]: Yes that's a good point.

I propose we do:
```
#[derive(Deserialize)]
struct CredentialPayload {
    provider: CredentialProvider,
    credentials: serde_json::Value,
}

File: core/bin/oauth.rs
Comment (spolu) [reply core/bin/oauth.rs:220]: Assuming the above this should just be payload.credentials

@Fraggle not that `raw_json` is used in `oauth` (and `core`) to name a string that contains JSON but has already been serialized.

File: core/src/oauth/credential.rs
Comment (spolu) [inline core/src/oauth/credential.rs:41]: should be called credentials

File: core/src/oauth/credential.rs
Comment (spolu) [inline core/src/oauth/credential.rs:49]: credentials. Also in the current code this is a string not a JSON which is misleading here but should be OK once we fix the API

Comment (Fraggle) [reply core/src/oauth/credential.rs:49]: Ok for naming.

On string vs json vs serialized json: 
If I understand correctly, we are converging toward expecting valid json (vs any string) but you also propose to use a real json format for the db column ?

However, don't we want to keep it encrypted ? meaning it has to stay a string, right ?

File: core/bin/oauth.rs
Comment (fontanierh) [reply core/bin/oauth.rs:210]: Still need a way to enforce it is a JSON object if we’re going for this (since any string is a serde value)

File: core/src/oauth/credential.rs
Comment (fontanierh) [reply core/src/oauth/credential.rs:49]: I dont think we want to use a json column in db

File: core/bin/oauth.rs
Comment (Fraggle) [reply core/bin/oauth.rs:220]: So to sum up:

- naming: `raw_json` => `credentials`
- payload expect JSON (using `serde_json::Value`
- then turned into an encrypted string in the DB (same as today)

Did I miss something ?

Comment (fontanierh) [reply core/bin/oauth.rs:220]: serde::Value won’t enforce json (since a simple string is a valid value)

Comment (Fraggle) [reply core/bin/oauth.rs:220]: You mean won't enforce json `object` right ?

Comment (Fraggle) [reply core/bin/oauth.rs:220]: Is `serde_json::Map<String, serde_json::Value>` the right type (I asked gpt4) ?

Comment (fontanierh) [reply core/bin/oauth.rs:220]: I think its right yes 👍

**Comment by Fraggle** [comment]
@fontanierh @spolu care to re-review ?

```
➜ curl -X POST http://localhost:3006/oauth-service/credentials \
     -H "Content-Type: application/json" \
     -d '{"provider": "snowflake", "credentials": {"hello": "world"}}' 

{
"error":null, 
"response": {
  "credential": 
   {
     "credential_id": "cred_5sQCZjos1M-f7a926d05a9c81dc765eef2219231c31a7585212258bb2a863c6635af30eede3",
     "created": 1726930106044, 
     "provider": "snowflake", 
     "credentials": {"hello":"world"}
   }
 }
}

➜  curl -X GET http://localhost:3006/oauth-service/credentials/cred_5sQCZjos1M-f7a926d05a9c81dc765eef2219231c31a7585212258bb2a863c6635af30eede3

{
"error":null, 
"response": {
  "credential": 
   {
     "credential_id": "cred_5sQCZjos1M-f7a926d05a9c81dc765eef2219231c31a7585212258bb2a863c6635af30eede3",
     "created": 1726930106044, 
     "provider": "snowflake", 
     "credentials": {"hello":"world"}
   }
 }
}
```

File: core/bin/oauth.rs
Comment (spolu) [inline core/bin/oauth.rs:210]: Let's do serde_json::Value this is the most flexible (any valid JSON)? It is also aligned with what we do for Connections.

File: core/bin/oauth.rs
Comment (spolu) [inline core/bin/oauth.rs:213]: please don't depart from the naming scheme of this file :pray: 

Should be `credentials_create`

File: core/bin/oauth.rs
Comment (spolu) [inline core/bin/oauth.rs:241]: Should be `credentials_finalize`

File: core/bin/oauth.rs
Comment (spolu) [inline core/bin/oauth.rs:310]: why `oauth-service` this is inconsistent with the existing endpoitns?

File: core/src/oauth/encryption.rs
Comment (spolu) [inline core/src/oauth/encryption.rs:20]: Given  you touch this code (even if no-op) can you add an explicit test to run before deploy.

Locally. Create a connection from `main` switch to your branch, retrieve the connection :pray: 

If sealing/unseal is inadvertently changed this would be pretty bad

File: core/bin/oauth.rs
Comment (spolu) [reply core/bin/oauth.rs:210]: This name is not ideal (I know I suggested it). Since the top-level object is called a `Credential` we should find something else as this is currently confusing.

content, value, secret ?

File: core/src/oauth/credential.rs
Comment (spolu) [inline core/src/oauth/credential.rs:41]: serde_json::Value

File: core/src/oauth/store.rs
Comment (spolu) [inline core/src/oauth/store.rs:243]: sealing/unsealing should be done outside of the store.rs file (if only out of consistency with the Connection code pathes + better division of concern)

File: core/src/oauth/credential.rs
Comment (spolu) [reply core/src/oauth/credential.rs:41]: Name TBD content, value, secret?

File: core/src/oauth/credential.rs
Comment (spolu) [inline core/src/oauth/credential.rs:42]: We should have encrypted_content/content/value here + sealing / unsealing should be done in that file not the store.

File: core/bin/oauth.rs
Comment (fontanierh) [reply core/bin/oauth.rs:210]: > Let's do serde_json::Value this is the most flexible (any valid JSON)? It is also aligned with what we do for Connections.

Disagree. If we do this, we get 0 value out of expecting json and we should just expect and store strings. With `serde_json::Value` you validate nothing because any string is a valid json value (i.e a string), so you don't prevent multiple levels of json.stringify.

in my view, it's either `String` (flexible) or `serde_json::Map<String, serde_json::Value>` (safer)

File: core/bin/oauth.rs
Comment (fontanierh) [reply core/bin/oauth.rs:241]: Why finalize ? There's nothing to finalize, so I don't understand the naming ? Should just be create / retrieve IMO.

File: core/bin/oauth.rs
Comment (fontanierh) [reply core/bin/oauth.rs:310]: should just be `/credentials` I think. Although I am still most bullish on calling the resource `secrets`

File: core/src/oauth/credential.rs
Comment (fontanierh) [reply core/src/oauth/credential.rs:41]: Disagree also

File: core/bin/oauth.rs
Comment (spolu) [reply core/bin/oauth.rs:210]: Then we want to update metadata in Connections

File: core/bin/oauth.rs
Comment (spolu) [reply core/bin/oauth.rs:241]: Sorry not finalize `credentials_retrieve`

File: core/bin/oauth.rs
Comment (Fraggle) [reply core/bin/oauth.rs:310]: oauth-service has probably been introduced by copilot, i overlooked it :)

File: core/src/oauth/credential.rs
Comment (Fraggle) [reply core/src/oauth/credential.rs:42]: Not against it at all but can you explain your reasonning ?

I felt it made sense to have the encryption / decryption to / from storage be done in the store file. We read data from the db and we return the content, the fact that it's encrypted in the db is a detail that the rest of the code do not need to be aware of, ihmo.

However, I see the inconsistency with connexion. I can move both to the "new pattern" if you agree that it's better.

**Comment by Fraggle** [comment]
@spolu @fontanierh I will fix the methods / endpoints on monday.

I'll let you guys decide wether you want any valid json value or absolutely a json object  :)

Personnally i'm leaning towards a json object & "credentials" (vs "secrets") because I think the service should not be too generic / too agnostic on it's purpose and content.

For example I can see ourselves adding some verifications on the credentials format based on the provider selected.

But I understand this is not necessary aligned with the generic k/v store that @fontanierh imagined in the first place.

File: core/src/oauth/encryption.rs
Comment (Fraggle) [reply core/src/oauth/encryption.rs:20]: > can you add an explicit test to run before deploy

I'm not sure what you mean:
- Me doing a manual test ?
- Unit test somehow ?
- Another form of test that I'm not aware of ?

File: core/bin/oauth.rs
Comment (Fraggle) [reply core/bin/oauth.rs:213]: Unintended, copilot wrote it and I overlook that it wasn't consistent.
Will fix.

**Comment by fontanierh** [comment]
> I'll let you guys decide wether you want any valid json value or absolutely a json object :)

I really am fine with either `String` or `serde::Map<...>`. There are advantages on both side IMO. 
What I am really against, however, is using `serde::Value`, which adds nothing vs `String` and makes everything more confusing (reading it, you might wrongly assume that the code is enforcing a JSON object)

>  I think the service should not be too generic / too agnostic on it's purpose and content

Personally I think this new "service" should not be over-fitting the remote databases use case. Right now, we use core's database to store `Providers` (eg  if a user adds their OpenAI key to use in dust apps), but IMO it wouldn't be crazy (I would even dare to say "principled") to migrate those to `oauth` secrets / credentlals in the future. 

That being said I think your point makes sense, storing the "provider" type and having some provider-specific validation code is not crazy either (since we already do that for `connections`). "provider-aware encrypted key/value store" doesn't sound bad. I'm still in favor of "dumb encrypted key/value store", but it's loosely held and I trust your judgement on this 👍

File: core/src/oauth/credential.rs
Comment (spolu) [reply core/src/oauth/credential.rs:42]: Main arguments are:
- separation of concern (having the crypto code not mixed with DB management)
- concistency with the rest of the service codebase

Comment (spolu) [reply core/src/oauth/credential.rs:42]: I think encryption should be in a very visible and contained place. store.rs does not really fit that IMHO

Comment (Fraggle) [reply core/src/oauth/credential.rs:42]: I updated the code to match what I think you had in mind. But now the unseal (the action not the actual code) happens in oauth.rs before sending the response which feel weird. Did I misunderstood ?

**Comment by Fraggle** [comment]
@fontanierh @spolu 

Pushed a revised PR with the discussed changes:
- `credentials` => `content`
- seal / unseal are done within `Credentials`
- kept "object" for the expected json and added some rudimentary checks for Snowflake
  - We control the insertion / deletion so we can decide on a format based on the provider, it should be easy to add OpenAi for example.

File: core/bin/oauth.rs
Comment (fontanierh) [inline core/bin/oauth.rs:240]: do we need to return the clear content from this endpoint ?

File: core/src/oauth/encryption.rs
Comment (Fraggle) [reply core/src/oauth/encryption.rs:20]: @spolu ping in case you missed it.

**Review by fontanierh — APPROVED** [review]
LGTM

File: core/bin/oauth.rs
Comment (Fraggle) [reply core/bin/oauth.rs:240]: not necessarely, removing.

File: core/src/oauth/credential.rs
Comment (fontanierh) [inline core/src/oauth/credential.rs:97]: I forgot, we also need `account` here

Comment (Fraggle) [reply core/src/oauth/credential.rs]: Done :)

**Comment by Fraggle** [comment]
@fontanierh I added metadata in the last commit PTAL but I'll not merge immediately.
Will do the refactoring to add tests in another PR, then rebase this one and make sure the tests don't break.

**Comment by spolu** [comment]
@Fraggle you probably already have this but I believe some work streams are blocked on this PR, as such there is definitely value in moving forward earlier than later and not blocking on more advanced testing (and do them after vs before :+1:)

**Pull request merged**
